File: SpiritSpenderClient/src/app/setup/types/drive-setting.ts
import { UnitsType } from './units-type';

export interface DriveSetting {
    driveName: string;
    stepsPerRevolution: number;
    spindlePitch: UnitsType;
    maxSpeed: UnitsType;
    acceleration: UnitsType;
    softwareLimitMinus: UnitsType;
    softwareLimitPlus: UnitsType;
    reverseDirection: boolean;
    referencePosition: UnitsType;
    referenceDrivingDirection: DrivingDirection;
    referenceDrivingSpeed: UnitsType;
}

export enum DrivingDirection{
    positive = 0,
    negative = 1
}
export class DriveSettingTexts {
    static readonly driveName = "Drive name";
    static readonly stepsPerRevolution = "Steps per Revolution";
    static readonly spindlePitch = "Spindle pitch";
    static readonly maxSpeed = "Max. speed";
    static readonly acceleration = "Acceleration";
    static readonly softwareLimitMinus = "Software limit -"
    static readonly softwareLimitPlus ="Software limit +"
    static readonly reverseDirection = "Reverse direction";
    static readonly referencePosition = "Reference position";
    static readonly referenceDrivingDirection = "Reference driving direction";
    static readonly referenceDrivingSpeed = "Reference driving speed";
}

File: SpiritSpenderServer/src/SpiritSpenderServer/Persistence/DriveSettings/DriveSetting.cs
namespace SpiritSpenderServer.Persistence.DriveSettings;

using MongoDB.Bson.Serialization.Attributes;
using Newtonsoft.Json;
using UnitsNet;

public class DriveSetting
{
    [BsonId]
    [JsonProperty]
    public string? DriveName { get; set; }
    [JsonProperty]
    public int StepsPerRevolution { get; set; }
    [JsonProperty]
    public Length SpindlePitch { get; set; }
    [JsonProperty]
    public Speed MaxSpeed { get; set; }
    [JsonProperty]
    public Acceleration Acceleration { get; set; }
    [JsonProperty]
    public Length SoftwareLimitMinus { get; set; }
    [JsonProperty]
    public Length SoftwareLimitPlus { get; set; }
    [JsonProperty]
    public bool ReverseDirection { get; set; }
    [JsonProperty]
    public Length ReferencePosition { get; set; }
    [JsonProperty]
    public DrivingDirection ReferenceDrivingDirection { get; set; }
    [JsonProperty]
    public Speed ReferenceDrivingSpeed { get; set; }
}

public enum DrivingDirection
{
    Positive = 0,
    Negative = 1
}
